### PR TITLE
Fix conflict error handling

### PR DESCRIPTION
**Pull request by alexshtin**
**What changed?**
- Fix few bugs in CAS failed operation error logic.
- Improve error messages. 
- Add CAS error logging.
- Add comments to clarify error handling logic.
- Rename generic var and param names to be more specific.

**Why?**
N/A

**How did you test it?**
Existing tests.

**Potential risks**
No risks.

**Is hotfix candidate?**
No.

### CODE REVIEW

File: common/persistence/cassandra/errors.go [hunk 1]
```diff
@@ -57,24 +57,17 @@ type (
 )
 
 func convertErrors(
-	record map[string]interface{},
```
Comment (wxing1292) [inline common/persistence/cassandra/errors.go:60]: why the original record is removed?

according to [MapExecuteBatchCAS](https://github.com/gocql/gocql/blob/769848eae4625444c6abdabc4a67eacb117c9200/session.go#L736-L750)
record will be filled with data if CAS failed

File: service/history/workflow/transaction_impl.go [hunk 2]
```diff
@@ -727,6 +730,10 @@ func emitCompletionMetrics(
 }
 
 func operationPossiblySucceeded(err error) bool {
+	if err == consts.ErrConflict {
+		return false
+	}
```
Comment (wxing1292) [inline service/history/workflow/transaction_impl.go:735]: i believe the logic should get rid of `consts.ErrConflict` in the long term

File: service/history/workflow/transaction_impl.go [hunk 3]
```diff
@@ -472,41 +472,44 @@ func updateWorkflowExecutionWithRetry(
...
+		switch err.(type) {
+		case *persistence.CurrentWorkflowConditionFailedError,
+			*persistence.WorkflowConditionFailedError,
+			*persistence.ConditionFailedError:
+			// TODO get rid of ErrConflict
+			return nil, consts.ErrConflict
+		default:
+			return nil, err
+		}
 	}
```
Comment (wxing1292) [inline service/history/workflow/transaction_impl.go:493]: if the intention is to emit error logs whenever error is encountered, then i would suggest log when error is generated

**Comment by yiminc** [comment]
Please update the PR description to include some context.

File: service/history/workflow/transaction_impl.go (hunk 2 above)
Comment (alexshtin) [reply service/history/workflow/transaction_impl.go:735]: There is a TODO for this (not by me), but I am not ready for this. Yet.

File: service/history/workflow/transaction_impl.go (hunk 3 above)
Comment (alexshtin) [reply service/history/workflow/transaction_impl.go:493]: I follow the rule: "Error should be propagated or handled. Not both". It gets propagated to this point and this substitution is sort of handling and good place to log it. I know we lost error call stack in this case.

File: common/persistence/cassandra/errors.go [hunk 4]
```diff
@@ -57,52 +57,62 @@ type (
...
 
-	record = make(map[string]interface{})
-	for iter.MapScan(record) {
-		records = append(records, record)
+	conflictRecord = make(map[string]interface{})
+	for conflictIter.MapScan(conflictRecord) {
+		if conflictRecord["[applied]"].(bool) {
+			// Should never happen. All records in batch should have [applied]=false.
+			continue
+		}
```
Comment (wxing1292) [inline common/persistence/cassandra/errors.go:82]: i guess this additional check may not be necessary

**Review by wxing1292 — APPROVED** [review]

**Pull request merged**